### PR TITLE
overwrite options

### PR DESCRIPTION
**Pull request by ocefpaf**
Implements https://github.com/ODM2/WOFpy/issues/139#issuecomment-307894647

There is a complicated logic there and I would like to look at it again tomorrow with fresh eyes before this is merged. Submitting the PR today just to get some input from @emiliom and @lsetiawan.

### CODE REVIEW

File: wof/wofpy_config.py [hunk 1]
```diff
@@ -16,62 +16,82 @@
 
 Usage:
     wofpy_config INDIR
-    wofpy_config INDIR [--mode=<development,production>]
+    wofpy_config INDIR [--mode=<development,production>] [--overwrite=<soft,hard,none>]
```
Comment (ocefpaf) [inline wof/wofpy_config.py]: Maybe we could make `soft` the default and drop the extra complexity of `none`. Experienced user would not expect to get things added into an existing directory, but if we write that in the docs it should be OK b/c there won't be loos of data.

File: wof/wofpy_config.py [hunk 2]
```diff
@@ -16,62 +16,82 @@
...
+  directory               Configuration directory.
 
 Options:
-  -h --help     Show this screen.
-  -v --version  Show version.
-  --mode=development   deploy mode [default: development]
+  -h --help               Show this screen.
+  -v --version            Show version.
+  -m --mode=development   Deploympent mode [default: development].
+  -o --overwrite=none     Overwrite everything *hard*,
```
Comment (ocefpaf) [inline wof/wofpy_config.py]: @lsetiawan I also added the short versions of the options. You can call `wofpy_config wofpyserver --mode=production --overwrite=soft` or `wofpy_config wofpyserver -m production -o soft` now.

File: wof/wofpy_config.py [hunk 3]
```diff
@@ -16,62 +16,82 @@
...
 _ODM2_TIMESERIES = os.path.join(_ROOT, 'examples', 'flask', 'odm2', 'timeseries')
 
 
-def makedirs(directory):
+def makedirs(directory, overwrite=None):
     if not os.path.exists(directory):
         os.makedirs(directory)
+    elif os.path.exists(directory) and overwrite in ['soft', 'hard']:
+        if overwrite == 'hard':
+            shutil.rmtree(directory)
```
Comment (ocefpaf) [inline wof/wofpy_config.py]: Maybe we should add a `[y/N]` before the deletion?

File: wof/wofpy_config.py [hunk 4]
```diff
@@ -16,62 +16,82 @@
...
 
-def copytree(src, dst, symlinks=False, ignore=None):
+def copytree(src, dst, symlinks=False, ignore=None, overwrite='none'):
     for item in os.listdir(src):
         if '.pyc' in item:
             continue
         s = os.path.join(src, item)
         d = os.path.join(dst, item)
+        if os.path.exists(d) and overwrite == 'soft':
+            continue
```
Comment (ocefpaf) [inline wof/wofpy_config.py]: Here we leave anu existing file untouch. This should not modify the user existing configuration.

File: wof/wofpy_config.py [hunk 5]
```diff
@@ -16,62 +16,82 @@
...
+    mode = args.get('--mode').lower()
     if mode not in ['development', 'production']:
         raise ValueError('Got mode: {!r}, expected development, or production.'.format(mode))
 
-    makedirs(directory)
+    overwrite = args.get('--overwrite').lower()
+    if overwrite not in ['hard', 'soft', 'none']:
+        msg = 'Got overwrite: {!r}, expected none, hard, or soft.'.format
+        raise ValueError(msg(overwrite))
+    elif overwrite == 'hard':
```
Comment (ocefpaf) [inline wof/wofpy_config.py]: We don't have to be this verbose by default. I can either implement a logging scheme or a `--verbose` option.

File: wof/wofpy_config.py (hunk 1 above)
Comment (lsetiawan) [inline wof/wofpy_config.py]: Agreed. I like the `soft` as the default. Thanks.

File: wof/wofpy_config.py (hunk 2 above)
Comment (lsetiawan) [inline wof/wofpy_config.py]: 😄  Woo hoo! You're awesome!

File: wof/wofpy_config.py (hunk 3 above)
Comment (lsetiawan) [inline wof/wofpy_config.py]: > It would block its use on scripts though.

What do you mean by this?

File: wof/wofpy_config.py (hunk 4 above)
Comment (lsetiawan) [inline wof/wofpy_config.py]: Yepp! This is exactly what I'm looking for 😃

File: wof/wofpy_config.py (hunk 5 above)
Comment (lsetiawan) [inline wof/wofpy_config.py]: @ocefpaf I don't think it is bad to be this verbose by default. At least for me, I'd like to know the options that the command is looking for.

File: wof/wofpy_config.py (hunk 2 above)
Comment (ocefpaf) [inline wof/wofpy_config.py]: `docopt` is quite powerful to create CLI. I prefer it to `click` or `fire` b/c it is only a POSIX parser on the `__doc__` text.

File: wof/wofpy_config.py (hunk 3 above)
Comment (ocefpaf) [inline wof/wofpy_config.py]: I mean that you cannot call it programatically when there is a user expected input.
Either way I am not convienced that is a good idea...

File: wof/wofpy_config.py (hunk 2 above)
Comment (lsetiawan) [inline wof/wofpy_config.py]: It seems like you can really customize the CLI Documentation as long as you know the standard way to write them. I thought docopt gets super complicated in a really complex CLI? I'd love to discuss your reasons choosing `click` over `docopt` at some point.

File: wof/wofpy_config.py (hunk 3 above)
Comment (lsetiawan) [inline wof/wofpy_config.py]: Ah yes. That makes sense. Just leave it be then.

**Comment by lsetiawan** [comment]
>There is a complicated logic there and I would like to look at it again tomorrow with fresh eyes before this is merged. Submitting the PR today just to get some input from @emiliom and @lsetiawan.

@ocefpaf, So are you still waiting till tomorrow or should I merge?

File: wof/wofpy_config.py (hunk 2 above)
Comment (ocefpaf) [inline wof/wofpy_config.py]: The `click` vs `docopt` discussion is almost the same as `argparse` vs `optparse`.
a POSIX only parser seems limited but it is actually quite powerful and field tested, after all virtually all of Unix CLIs as like that.

Both `click` and `argarse` will allow you to create fancier CLI, with complex options and defaults, but the real question is: should you be creating a CLI like that :stuck_out_tongue_winking_eye:

On the other hand `click` does support subcommands while `docopt` would rely on pipes to get a similar functionality.

TL;DR it is a matter of taste but I like to be on the side of simplicity and easy manitainability.
So unless I am building a CLI tool that will have many subcommands, like `conda (-build,-env,-install,-create,etc)`, I prefer `optparse`, or its easier to use cousin `docopt`.

**Comment by ocefpaf** [comment]
> @ocefpaf, So are you still waiting till tomorrow or should I merge?

Let me push one extra commit and you can merge and test it locally. We can fix anything you find in another PR then.

**Comment by emiliom** [comment]
Nice! Thanks to both.
